components/region_cache_memory_engine/src/memory_controller.rs

```diff
@@ -0,0 +1,140 @@
+// Copyright 2024 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+
+use crate::{engine::SkiplistEngine, write_batch::NODE_OVERHEAD_SIZE_EXPECTATION};
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum MemoryUsage {
+    NormalUsage(usize),
+    SoftLimitReached(usize),
+    // usize here means the current memory usage and it's the usize in it adding with the memory
+    // acquiring exceeds the hard limit
+    HardLimitReached(usize),
+}
+
+/// MemoryController is used to control the memory usage of the region cache
+/// memory engine. The memory consumption is calculated by the allocated memory
+/// for keys and values, and the overhead of the skiplist where the overhead is
+/// estimated by using the node count of the skiplist times the
+/// `NODE_OVERHEAD_SIZE_EXPECTATION`.
+#[derive(Debug)]
+pub struct MemoryController {
+    // Allocated memory for keys and values (node overhead is not included)
+    allocated: AtomicUsize,
+    soft_limit_threshold: usize,
+    hard_limit_threshold: usize,
+    memory_checking: AtomicBool,
+
+    // The number of writes that are buffered but not yet written.
+    pending_node_count: AtomicUsize,
+    skiplist_engine: SkiplistEngine,
```

Comment:
Some general thoughts, should we split it in two classes? One class to control memory usage of keys, values, batch entries, and another (in a hierarchy?) to control memory usage of the skiplist itself. This could be an interface. I've worked on a similar _exact_ memory tracker a while ago, perhaps this could apply: https://github.com/apache/kudu/blob/master/src/kudu/util/mem_tracker.h (feel free to disregard this suggestion, however.) Thoughts, @tonyxuqqi?

Reply:
What's the benefit of split? Here, the skiplist has nothing to controller. It just return the number of nodes so that we use it the estimate the skiplist overhead memory usage.

Reply:
For this one, I think it's ok to merge them together for now. I cannot think of a case when we need to have separate data. But we should be able to get the kv count in metrics
@SpadeA-Tang